dm/pkg/shardddl/optimism/lock.go

```diff
@@ -871,3 +714,419 @@ func GetColumnName(lockID, ddl string, tp ast.AlterTableType) (string, error) {
 	}
 	return "", nil
 }
+
+// checkAddDropColumn check for ALTER TABLE ADD/DROP COLUMN statement
+// FOR ADD COLUMN, check whether add column with a different field or add a droped column
+// FOR DROP COLUMN, return the droped column.
+func (l *Lock) checkAddDropColumn(source, schema, table string, ddl string, prevTable, postTable schemacmp.Table) (string, error) {
+	currTable := l.tables[source][schema][table]
+	defer func() {
+		l.tables[source][schema][table] = currTable
+	}()
+
+	l.tables[source][schema][table] = prevTable
+	oldJoined, err := l.joinNormalTables()
+	if err != nil {
+		// nolint:nilerr
+		return "", nil
+	}
+
+	l.tables[source][schema][table] = postTable
+	newJoined, err := l.joinNormalTables()
+	if err != nil {
+		// nolint:nilerr
+		return "", nil
+	}
+
+	cmp, err := oldJoined.Compare(newJoined)
+	if err != nil {
+		// nolint:nilerr
+		return "", nil
+	}
+
+	if cmp <= 0 {
+		if col, err2 := AddDifferentFieldLenColumns(l.ID, ddl, oldJoined, newJoined); err2 != nil {
+			// check for add column with a larger field len
+			return "", err2
+		} else if _, err2 = AddDifferentFieldLenColumns(l.ID, ddl, postTable, newJoined); err2 != nil {
+			// check for add column with a smaller field len
+			return "", err2
+		} else if len(col) > 0 && l.IsDroppedColumn(source, schema, table, col) {
+			return "", terror.ErrShardDDLOptimismTrySyncFail.Generate(l.ID, fmt.Sprintf("add column %s that wasn't fully dropped in downstream. ddl: %s", col, ddl))
+		}
+	}
+
+	if cmp >= 0 {
```

Comment:
why we both check `if cmp <=0` and `if cmp >=0`🤔

Reply:
``
source1: add column a varchar(4) -- info1
source2: add column a varchar(5) -- info2
```
when receive info2, cmp will be 0, we need check for error

```
source1: drop column a -- info1
source2: drop column a -- info2
```
when reveice info1, cmp will be 0, we need check for drop column